File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/MyShopDsl.java
package com.mycompany.myshop.testkit.dsl.core.usecase;

import com.mycompany.myshop.testkit.driver.port.MyShopDriver;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.BrowseCoupons;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.CancelOrder;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.DeliverOrder;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.GoToMyShop;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.PlaceOrder;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.PublishCoupon;
import com.mycompany.myshop.testkit.dsl.core.usecase.usecases.ViewOrder;
import com.mycompany.myshop.testkit.common.Closer;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;

import java.io.Closeable;

public class MyShopDsl implements Closeable {
    private final MyShopDriver driver;
    private final UseCaseContext context;

    public MyShopDsl(MyShopDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToMyShop goToMyShop() {
        return new GoToMyShop(driver, context);
    }

    public PlaceOrder placeOrder() {
        return new PlaceOrder(driver, context);
    }

    public CancelOrder cancelOrder() {
        return new CancelOrder(driver, context);
    }

    public DeliverOrder deliverOrder() {
        return new DeliverOrder(driver, context);
    }

    public ViewOrder viewOrder() {
        return new ViewOrder(driver, context);
    }

    public PublishCoupon publishCoupon() {
        return new PublishCoupon(driver, context);
    }

    public BrowseCoupons browseCoupons() {
        return new BrowseCoupons(driver, context);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/MyShopDsl.ts
import type { MyShopDriver } from '../../../driver/port/my-shop-driver.js';
import type { UseCaseContext } from '../shared/use-case-context.js';
import { PlaceOrder } from './usecases/PlaceOrder.js';
import { ViewOrder } from './usecases/ViewOrder.js';
import { CancelOrder } from './usecases/CancelOrder.js';
import { DeliverOrder } from './usecases/DeliverOrder.js';
import { BrowseCoupons } from './usecases/BrowseCoupons.js';
import { PublishCoupon } from './usecases/PublishCoupon.js';
import { GoToMyShop } from './usecases/GoToMyShop.js';

export class MyShopDsl {
  constructor(
    private readonly driver: MyShopDriver,
    private readonly context: UseCaseContext,
  ) {}

  goToMyShop(): GoToMyShop {
    return new GoToMyShop(this.driver, this.context);
  }

  placeOrder(): PlaceOrder {
    return new PlaceOrder(this.driver, this.context);
  }

  viewOrder(): ViewOrder {
    return new ViewOrder(this.driver, this.context);
  }

  cancelOrder(): CancelOrder {
    return new CancelOrder(this.driver, this.context);
  }

  deliverOrder(): DeliverOrder {
    return new DeliverOrder(this.driver, this.context);
  }

  browseCoupons(): BrowseCoupons {
    return new BrowseCoupons(this.driver, this.context);
  }

  publishCoupon(): PublishCoupon {
    return new PublishCoupon(this.driver, this.context);
  }

  async close(): Promise<void> {
    await this.driver.close();
  }
}
